[LoopVectorize] Use predicated version of getSmallConstantMaxTripCount

david-arm opened the pull request:
There are a number of places where we call getSmallConstantMaxTripCount
without passing a vector of predicates:

getSmallBestKnownTC
isIndvarOverflowCheckKnownFalse
computeMaxVF
isMoreProfitable

I've changed all of these to now pass in a predicate vector so that
we get the benefit of making better vectorisation choices when we
know the max trip count for loops that require SCEV predicate checks.

I've tried to add tests that cover all the cases affected by these
changes.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -3994,8 +3996,13 @@ LoopVectorizationCostModel::computeMaxVF(ElementCount UserVF, unsigned UserIC) {
+
+  SmallVector<const SCEVPredicate *, 2> Predicates;
+  unsigned MaxTC =
+      PSE.getSE()->getSmallConstantMaxTripCount(TheLoop, &Predicates);
```
Should this use PSE.getSmallConstantMaxTripCount() to ensure the predicates are added to PSE?

david-arm replied:
Hmm, that's a good question. I'd assumed that when we call PSE.getBackedgeTakenCount(TheLoop, &Predicates) at some point later on to create the trip count, that it would just end up adding the same predicates anyway.

Does it matter if we attempt to add the same predicate twice? If not, then I'm happy to change this and use the predicated version.

david-arm replied:
I think it's safe to add the predicates multiple times, since the `addPredicate` function in PredicatedScalarEvolution already checks to see if the new predicate is already implied by its existing list. I've added a new `getSmallConstantMaxTripCount` to PredicatedScalarEvolution and kept track of the count to avoid unnecessary computation, similar to how we return the backedge taken count.

fhahn replied:
Yes, PSE should take care to avoid adding duplicated predicates, thanks for adjusting!

fhahn replied:
Should a similar change also be done for the recent changes for early exit checks in LV?

david-arm replied:
Do you mean the calls to `PSE.getSE()->getPredicatedExitCount` in `LoopVectorizationLegality::isVectorizableEarlyExitLoop`? I'm not sure if we want to do exactly the same, i.e. cache the exit count for every exit. I could potentially add an equivalent method to `PredicatedScalarEvolution` that automatically adds the predicates, but probably best done in a separate patch because it's unrelated to the max trip count. What do you think?

david-arm commented:
Deal with conflicts are rebase

fhahn commented on llvm/include/llvm/Analysis/ScalarEvolution.h:
```diff
@@ -2447,6 +2451,9 @@ class PredicatedScalarEvolution {
   const SCEV *SymbolicMaxBackedgeCount = nullptr;
+
+  /// The constant max trip count for the loop.
+  std::optional<unsigned> SmallConstantMaxTripCount;
```
```suggestion
const SCEV* SmallConstantMaxTripCount = nullptr;
```

for consistency with `BackedgeCount` and `SymbolicMaxBackedgeCount` above?

david-arm replied:
The problem with that is the max trip count is measured as an `unsigned` value. I could potentially cache the intermediate result in the function:

```
unsigned ScalarEvolution::getSmallConstantMaxTripCount(
    const Loop *L, SmallVectorImpl<const SCEVPredicate *> *Predicates) {

  const auto *MaxExitCount =
      Predicates ? getPredicatedConstantMaxBackedgeTakenCount(L, *Predicates)
                 : getConstantMaxBackedgeTakenCount(L);
  return getConstantTripCount(dyn_cast<SCEVConstant>(MaxExitCount));
}
```

i.e. `MaxExitCount`, but then you still have to call `getConstantTripCount` each time so you see less benefit from caching it. That's why I chose to use `std::optional<unsigned>` - alternatively I could use a larger type such as `uint64_t` and treat UINT64_MAX as being equivalent to `not cached yet`.

fhahn replied:
Ah  I see! It's not worth the trouble then I think.

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/low_trip_count_predicates.ll:
```diff
@@ -0,0 +1,405 @@
+  br label %while.body
+
+while.body:
+  %indvars.iv = phi i64 [ %0, %while.preheader ], [ %indvars.iv.next, %while.body ]
```
nit: for shorter value names

```suggestion
  %iv = phi i64 [ %0, %while.preheader ], [ %indvars.iv.next, %while.body ]
```

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/low_trip_count_predicates.ll:
```diff
@@ -0,0 +1,405 @@
+; CHECK-NEXT:    ret void
+;
+entry:
+  %p.promoted = load i32, ptr %p, align 4
```
simpler to pass as arg?

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/low_trip_count_predicates.ll:
```diff
@@ -0,0 +1,405 @@
+while.body:
+  %indvars.iv = phi i64 [ %0, %while.preheader ], [ %indvars.iv.next, %while.body ]
+  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
+  %arrayidx = getelementptr inbounds nuw [100 x i8], ptr %v, i64 0, i64 %indvars.iv
```
nit: slightly simpler to use

```suggestion
  %arrayidx = getelementptr inbounds nuw i8, ptr %v, i64 %indvars.iv
```

david-arm committed:
Add tests and extra max trip count debug output

david-arm committed:
[LoopVectorize] Use predicated version of getSmallConstantMaxTripCount

There are a number of places where we call getSmallConstantMaxTripCount
without passing a vector of predicates:

getSmallBestKnownTC
isIndvarOverflowCheckKnownFalse
computeMaxVF
isMoreProfitable

I've changed all of these to now pass in a predicate vector so that
we get the benefit of making better vectorisation choices when we
know the max trip count for loops that require SCEV predicate checks.

I've tried to add tests that cover all the cases affected by these
changes.

fhahn reviewed (approved):
LGTM, thanks!
